### PR TITLE
WIP: Bootstrap 4 upgrade

### PR DESCRIPTION
**Pull request by laser505t**
There are still some things I need to do:
- [ ] Get ood_appkit out of versioning once I've created a PR for that
- [ ] Be more exact about the gem version requirements

Other than those items I think the only things to discuss are design decisions, both in ood_appkit and in the pseudofun app itself.

### CODE REVIEW

**Comment by laser505t** [comment]
Old:
image

New:
image

The status badges not working are because of osc_machete_rails not being compatible with Bootstrap 4, which I opened an issue in last week.

File: Gemfile [hunk 1]
```diff
@@ -42,8 +42,8 @@ end
 
 
 gem 'autoprefixer-rails'
-gem 'bootstrap-sass', '~> 3.3'
-gem 'bootstrap_form', '~> 2.5'
+gem 'bootstrap'
+gem 'bootstrap_form'
```
Comment (ericfranz) [inline Gemfile:46]: Add back version qualifier but to latest required i.e. `~> 2.7` if that is the required version...

File: app/views/blasts/index.html.erb [hunk 2]
```diff
@@ -1,4 +1,4 @@
-<div class="page-header">
+<div class="mt-5 mb-3 border-bottom">
```
Comment (ericfranz) [inline app/views/blasts/index.html.erb:1]: Should we consider adding a custom page-header class that uses sass to include `mt-5 mb-3 border-bottom`?

**Comment by ericfranz** [comment]
- [x] height of the navbar? can this be fixed to be same as the original? what would be involved?
- [x] `<pre>` is missing the "well" look
- [ ] bootstrap_form problem? validation errors don't highlight the form elements = update bootstrap_form?
- [x] font size change? is there an explanation of this choice somewhere?
- [ ] crimson color is off (change it back?? maybe...) - is there a new theme mechanism? like what would we do if we wanted the original bootstrap 3 color scheme in bootstrap 4?
- [ ] Add a `.page-header` and `.btn-default { @include .btn-light }` class maybe (alt branch)
- [x] Try adding a modified helper for the status label and see if that overrides the osc_machete_rails gem's helper. If necessary, use the full namespace for this gem to execute the correct helper.

**Comment by laser505t** [comment]
Unfortunately the bootstrap_form version 4.0.0 gem requires Rails 5 as a runtime dependency, which seems weird because the version we used, 2.7.0, has no runtime dependencies but a rather long list of dev dependencies that include Rails 4.

**Comment by laser505t** [comment]
From Bootstrap 4 migration documentation: https://getbootstrap.com/docs/4.0/migration/
```
Global font-size increased from 14px to 16px.
```
Technically its really 1rem, but apparently most browsers default that to be 16px.
Also, fonts are intended to match the native fonts of the device which might have some effect: https://getbootstrap.com/docs/4.0/content/reboot/#native-font-stack

**Comment by ericfranz** [comment]
Okay, so we can assume that upgrading to Rails 5 is a requirement for upgrading to Bootstrap 4, and that we should do the Rails 5 upgrade first.

**Comment by ericfranz** [comment]
If we do that we will likely look at upgrading to Rails 5.0.0 (or whatever the exact version is of https://www.softwarecollections.org/en/scls/rhscl/rh-ror50/).

I'm okay if you want to upgrade pseudofun to Rails 5. You could do that in a separate PR request and then we could rebase your changes in this PR on top of the Rails 5 - I don't think they would conflict.

Follow https://guides.rubyonrails.org/v5.0/upgrading_ruby_on_rails.html exactly. Upgrade to latest patch version of Rails 4.2 first. Then follow the directions for upgrading to Rails 5.0.0. Besides updating the gem file and modifying files manually, there are rake tasks to help with updating an app.

**Comment by ericfranz** [comment]
Once you have upgraded to bootstrap 4 gem, it should be straight forward to use the sass includes from that gem for customizing bootstrap just like we did for bootstrap 3. That will give us the same color format, and we can probably customize other things if we want like the default navbar size etc.

**Pull request closed without merging**